src/p11_wrap.cpp

```diff
@@ -0,0 +1,283 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+/*
+ * This file is the only place where we should see any
+ * methods from LibP11. Any other code should use
+ * the more high-level methods declared and define in
+ * p11_wrap.cpp and p11_wrap.h.
+ *
+ */
+
+#include <boost/algorithm/hex.hpp>
+#include <cassert>
+#include <cstddef> /* this has to come before cppc (bug in boost) */
+#include <exception>
+
+#include <cppc/checkcall.hpp>
+#include "mococrw/openssl_wrap.h"
+#include "mococrw/p11_wrap.h"
+
+namespace mococrw
+{
+namespace p11
+{
+std::string P11Exception::generateP11ErrorString()
+{
+    // Note, pkcs11_CTX_new() internally loads LibP11 error strings,
+    // i.e., it invokes ERR_load_PKCS11_strings(). We therefore
+    // don't trigger the loading ourselves.
+    auto error = openssl::lib::OpenSSLLib::SSL_ERR_get_error();
+    auto formatter = boost::format("%s: %d");
+    formatter % openssl::lib::OpenSSLLib::SSL_ERR_error_string(error, nullptr) % error;
+    return formatter.str();
+}
+
+/**
+ * This struct is used by CWrap to
+ * determine how to react to an error
+ * when performing a LibP11 call.
+ */
+struct P11ExceptionErrorPolicy
+{
+    template <class Rv>
+    static void handleError(const Rv &);
+};
+
+/**
+ * Throw a P11 exception upon error.
+ *
+ * This method gets the error string and throws an exception
```

Comment:
"gets" is misleading here.
I guess the better description would be:
P11Exception generates the error string from the openssl error string in the constructor.
```suggestion
 * This method fetches the error string from openssl in P11Exception's constructor and throws an exception
```

Reply:
Done. I did s/gets/fetches but refrained from delving deeper into implementation details for abstraction.